provider/provider_config/spark_config.go

```diff
@@ -0,0 +1,217 @@
+	if err != nil {
+		return fmt.Errorf("unmarshal: %w", err)
+	}
+
+	s.ExecutorType = temp.ExecutorType
+	s.StoreType = temp.StoreType
+
+	err = s.decodeExecutor(temp.ExecutorType, temp.ExecutorConfig)
+	if err != nil {
+		return fmt.Errorf("could not decode executor: %w", err)
+	}
+
+	err = s.decodeFileStore(temp.StoreType, temp.StoreConfig)
+	if err != nil {
+		return fmt.Errorf("could not decode filestore: %w", err)
+	}
+
+	return nil
+}
+
+func (s SparkConfig) MutableFields() ss.StringSet {
+	result := ss.StringSet{}
+	var executor ss.StringSet
+	var store ss.StringSet
+
+	switch s.ExecutorType {
+	case EMR:
+		executor = s.ExecutorConfig.(*EMRConfig).MutableFields()
+	case Databricks:
+		executor = s.ExecutorConfig.(*DatabricksConfig).MutableFields()
+	}
+
+	switch s.StoreType {
+	case Azure:
+		store = s.StoreConfig.(*AzureFileStoreConfig).MutableFields()
+	case S3:
+		store = s.StoreConfig.(*S3FileStoreConfig).MutableFields()
+	case GCS:
+		store = s.StoreConfig.(*GCSFileStoreConfig).MutableFields()
+	}
+
+	for field, val := range executor {
+		result["Executor."+field] = val
+	}
+
+	for field, val := range store {
+		result["Store."+field] = val
+	}
+
+	return result
+}
+
+func (a SparkConfig) DifferingFields(b SparkConfig) (ss.StringSet, error) {
+	result := ss.StringSet{}
+	var executor ss.StringSet
+	var store ss.StringSet
+	var err error
+
+	switch a.ExecutorType {
+	case EMR:
+		if b.ExecutorType != EMR {
+			err = fmt.Errorf("executor config mismatch: a = %v; b = %v", a.ExecutorType, b.ExecutorType)
+		} else {
+			executor, err = a.ExecutorConfig.(*EMRConfig).DifferingFields(*b.ExecutorConfig.(*EMRConfig))
+		}
+	case Databricks:
+		if b.ExecutorType != Databricks {
+			err = fmt.Errorf("executor config mismatch: a = %v; b = %v", a.ExecutorType, b.ExecutorType)
+		} else {
+			executor, err = a.ExecutorConfig.(*DatabricksConfig).DifferingFields(*b.ExecutorConfig.(*DatabricksConfig))
+		}
+	}
+
+	if err != nil {
+		return result, err
+	}
+
+	switch a.StoreType {
+	case Azure:
+		if b.StoreType != Azure {
+			err = fmt.Errorf("store config mismatch: a = %v; b = %v", a.StoreType, b.StoreType)
+		} else {
+			store, err = a.StoreConfig.(*AzureFileStoreConfig).DifferingFields(*b.StoreConfig.(*AzureFileStoreConfig))
+		}
+	case S3:
+		if b.StoreType != S3 {
+			err = fmt.Errorf("store config mismatch: a = %v; b = %v", a.StoreType, b.StoreType)
+		} else {
+			store, err = a.StoreConfig.(*S3FileStoreConfig).DifferingFields(*b.StoreConfig.(*S3FileStoreConfig))
+		}
+	case GCS:
+		if b.StoreType != GCS {
+			err = fmt.Errorf("store config mismatch: a = %v; b = %v", a.StoreType, b.StoreType)
+		} else {
+			store, err = a.StoreConfig.(*GCSFileStoreConfig).DifferingFields(*b.StoreConfig.(*GCSFileStoreConfig))
+		}
+	}
+
+	if err != nil {
+		return result, err
+	}
+
+	for field, val := range executor {
+		result["Executor."+field] = val
+	}
+
+	for field, val := range store {
+		result["Store."+field] = val
+	}
+
+	return result, err
+}
```

Comment:
@sdreyer, @simba-git, and @ahmadnazeri, let me know if this matches your expectations for a valid spark config update.

Reply:
I think the logic makes sense